### PR TITLE
Add Hermes engine as option for macOS

### PR DESCRIPTION
**Pull request by alloy**
Related project board https://github.com/microsoft/react-native-macos/projects/4

#### Please select one of the following

- [ ] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [x] I am making a fix / change for the macOS implementation of react-native
- [ ] I am making a change required for Microsoft usage of react-native

## Summary

This adds the [Hermes engine](https://hermesengine.dev) as an optional replacement to JavaScriptCore for macOS targets.

* Hermes is enabled by virtue of adding the `React-Core/Hermes` dependency to one’s Podfile.

* The engine is shipped as a prebuilt framework binary, optimised for release.

* It comes with a separate debug symbols bundle (dSYM) for debugging/crash report symbolication.
  
   hermes-framework-loaded-with-dSYM

   availability-of-debug-symbols

* Support to generate source-maps when building release builds has been added, both without Hermes enabled:
   
   source-maps-jsc

   …and with Hermes enabled:

   source-maps-hermes

   (Note that file/line is correct, but column not. At the moment I am brushing this off as a minor issue, but something to watch out for, especially when we move to Hermes `master`.)

## Changelog

[macOS] [Added] - Add Hermes as (optional) JS engine.
[macOS/iOS] [Added] - Add ability to generate source-maps when building a release build with Xcode.

## Test Plan

* Update `Podfile` with `use_react_native!(hermes_enabled => true)` to add the dependencies and enable it as runtime engine.
* Uncomment `FORCE_BUNDLING` in bundle build-script phase to test source-maps.
* Verify that attaching a Chrome debugger works.

## TODO

- [x] Finish this PR: review it myself and finish the PR description.
- [x] Rebase on latest RN macOS v0.62.0 branch.
- [ ] The changes required to build the binary are currently still [here in my fork](https://github.com/alloy/hermes/tree/add-binary-podspec-for-macos-v0.4.1). I am reaching out to the upstream team about incorporating those changes upstream and releasing npm packages with the binary from their account, especially going forward.

    This could also be done in a follow-up PR, as long as we do so before we release v0.62.0 (preview); although in that case it might be preferable to move the tarball from my account to this MS repo?

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-macos/pull/473)

### CODE REVIEW

**Comment by alloy** [comment]
The current CI failures are identical to those on the `fb62merge` branch.

File: React/CxxBridge/RCTCxxBridge.mm [hunk 1]
```diff
@@ -347,7 +355,21 @@ - (void)start
       executorFactory = [cxxDelegate jsExecutorFactoryForBridge:self];
     }
     if (!executorFactory) {
-      executorFactory = std::make_shared<JSCExecutorFactory>(nullptr);
+      auto installBindings =
```
Comment (tom-un) [inline React/CxxBridge/RCTCxxBridge.mm:358]: Should this block be inside the #else (!RCT_USE_HERMES) ?

Comment (alloy) [reply React/CxxBridge/RCTCxxBridge.mm:358]: Both the `HermesExecutorFactory` and the `JSCExecutorFactory` need this block, which is why I moved it out of `JSCExecutorFactory.mm` into here.

Comment (tom-un) [reply React/CxxBridge/RCTCxxBridge.mm:358]: The previous implementation just did `executorFactory = std::make_shared<JSCExecutorFactory>(nullptr);`

I tested it locally: the same log events happen, i.e. _RCTLogJavaScriptInternal gets called for all logs, but they happen in a different callstack frame:

With `executorFactory = std::make_shared<JSCExecutorFactory>(installBindings);` the callstack for logs are:
```
#0	_RCTLogJavaScriptInternal
#1	-[RCTCxxBridge start]::$_0::operator()(facebook::jsi::Runtime&) const::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int) const 
#2	decltype(std::__1::forward<-[RCTCxxBridge start]::$_0::operator()(facebook::jsi::Runtime&) const::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)&>(fp)(std::__1::forward<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(fp0), std::__1::forward<unsigned int>(fp0))) std::__1::__invoke<-[RCTCxxBridge start]::$_0::operator()(facebook::jsi::Runtime&) const::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int>(-[RCTCxxBridge start]::$_0::operator()(facebook::jsi::Runtime&) const::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int&&) at /Applications/Xcode_11.5.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/type_traits:4425
...
``` 
With `executorFactory = std::make_shared<JSCExecutorFactory>(nullptr);` the callstack for logs are:
```
#0 _RCTLogJavaScriptInternal 
#1 facebook::react::JSCExecutorFactory::createJSExecutor(std::__1::shared_ptr<facebook::react::ExecutorDelegate>, std::__1::shared_ptr<facebook::react::MessageQueueThread>)::$_0::operator()(facebook::jsi::Runtime&) const::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int) const 
#2 decltype(std::__1::forward<facebook::react::JSCExecutorFactory::createJSExecutor(std::__1::shared_ptr<facebook::react::ExecutorDelegate>, std::__1::shared_ptr<facebook::react::MessageQueueThread>)::$_0::operator()(facebook::jsi::Runtime&) const::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)&>(fp)(std::__1::forward<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(fp0), std::__1::forward<unsigned int>(fp0))) std::__1::__invoke<facebook::react::JSCExecutorFactory::createJSExecutor(std::__1::shared_ptr<facebook::react::ExecutorDelegate>, std::__1::shared_ptr<facebook::react::MessageQueueThread>)::$_0::operator()(facebook::jsi::Runtime&) const::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int>(facebook::react::JSCExecutorFactory::createJSExecutor(std::__1::shared_ptr<facebook::react::ExecutorDelegate>, std::__1::shared_ptr<facebook::react::MessageQueueThread>)::$_0::operator()(facebook::jsi::Runtime&) const::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int&&) 
...
```
I don't think it matters.   I was just worried that we might be getting some new logging events, or duplicated logging events, but we're not: we're getting identical logging events, and in the same order, just from a slightly different callstack.  So I'm ok with it.

Comment (alloy) [reply React/CxxBridge/RCTCxxBridge.mm:358]: Thanks for being vigilant!

File: third-party-podspecs/Folly.podspec [hunk 2]
```diff
@@ -70,6 +70,40 @@ Pod::Spec.new do |spec|
...
+                           'folly/concurrency/*.{h,cpp}',
+                           'folly/system/{ThreadId,ThreadName}.{h,cpp}',
+                           'folly/synchronization/*.{h,cpp}',
+                           'folly/synchronization/detail/*.{h,cpp}',
+                           'folly/experimental/{ExecutionObserver,ReadMostlySharedPtr,TLRefCount}.{h,cpp}',
+                           'folly/io/async/{AsyncTimeout,DelayedDestruction,DelayedDestructionBase,EventBase,EventBaseManager,EventHandler,EventUtil,HHWheelTimer,NotificationQueue,Request,TimeoutManager,VirtualEventBase}.{h,cpp}',
+                           'folly/io/{Cursor,Cursor-inl,IOBuf,IOBufQueue}.{h,cpp}',
+                           'folly/tracing/StaticTracepoint.{h,cpp}',
+                           'folly/{Executor,ExceptionWrapper,ExceptionWrapper-inl,FileUtil,Singleton,SharedMutex}.{h,cpp}',
+                           'folly/detail/{AtFork,Futex,Futex-inl,MemoryIdler,StaticSingletonManager,ThreadLocalDetail}.{h,cpp}',
```
Comment (alloy) [inline third-party-podspecs/Folly.podspec:91]: For posterity sake, this was missing `folly/detail/AsyncTrace.*`

**Pull request merged**